lib/usart/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! USART interface.
+//!
+//! This was formerly a driver and the implementation still looks largely
+//! driver-like in that it interacts with other drivers and system interrupts,
+//! but USARTs are inherently single-owner. Structuring USARTs as a lib allows
+//! the calling task to use the USART more directly than if it were a
+//! full-fledged driver of its own.
+
+#![no_std]
+
+#[cfg(any(feature = "stm32h743", feature = "stm32h753"))]
+pub mod stm32h7;
+
+#[cfg(any(feature = "stm32h743", feature = "stm32h753"))]
+use self::stm32h7::Device;
+
+use tinyvec::ArrayVec;
+
+/// Supported baud rates
+#[derive(Debug, Clone, Copy)]
+#[repr(u32)]
+pub enum BaudRate {
+    Rate9600 = 9_600,
+    Rate57600 = 57_600,
+    Rate115200 = 115_200,
+}
+
+/// Handle to an enabled USART device.
+pub struct Usart<const TX_BUF_LEN: usize, const RX_BUF_LEN: usize> {
+    usart: Device,
+    irq_mask: u32,
+    tx_buf: ArrayVec<[u8; TX_BUF_LEN]>,
+    rx_buf: ArrayVec<[u8; RX_BUF_LEN]>,
+    rx_overrun: bool,
+}
+
+/// Errors detected during rx
+pub enum RxError {
+    /// Data has been lost due to rx buffer overrun
+    Overrun,
+}
+
+impl<const TX_BUF_LEN: usize, const RX_BUF_LEN: usize>
+    Usart<TX_BUF_LEN, RX_BUF_LEN>
+{
+    /// Start managing `device`.
+    ///
+    /// Enables the `irq` interrupt; the caller is responsible for calling
+    /// [`Usart::handle_interrupt()`] when that interrupt fires.
+    pub fn new(usart: Device, irq_mask: u32) -> Self {
+        // Turn on our interrupt. We haven't enabled any interrupt sources at
+        // the USART side yet, but we will momentarily.
+        userlib::sys_irq_control(irq_mask, true);
```

Comment:
(Nit: I'd probably leave this in control of the calling code, an operation like `new` turning on some interrupts might be surprising.)